modules/querier/external/token_provider.go

```diff
@@ -0,0 +1,73 @@
+package external
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"golang.org/x/oauth2"
+	"google.golang.org/api/idtoken"
+)
+
+// Caches an oauth2.TokenSource to enable efficient auth on each of our external
+// endpoints.
+type cachedTokenProvider struct {
+	tokenSources map[string]oauth2.TokenSource
+}
+
+func newTokenProvider(
+	ctx context.Context,
+	endpoints []string,
+	getTokenSource func(ctx context.Context, endpoint string) (oauth2.TokenSource, error),
```

Comment:
Yeah, I can go either way, but I went ahead and added this approach to see. There's a standalone commit that implements this change, if that's easier to review.

On one hand, this approach normalizes our various token providers under the same interface. From an OO perspective, I can see how it's easier to reason about.

On the other hand, we now have an extra `getTokenSource` function on the interface, which is required when wrapping the provider with caching. So we have to define the function on the cachedTokenProvider to satisfy the interface, even though it's not meant to be called.

I think either approach is reasonable, and it's fun thinking through these tradeoffs. But lmk if I missed something in your suggestion :)

Reply:
I thought more about this, and decided to revert my approach described above because I think the original is more maintainable. You can still review the change under the commit titled `implement same interface for all token providers` - perhaps I missed something in your suggestion that made the interface feel bloated. If you still think we should use the same interface, I'm always happy to chat and/or address it as a follow up. I'll continue thinking about this as well :thinking: